Source/JavaScriptCore/runtime/SetPrototype.cpp

```diff
@@ -175,6 +181,175 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncSize, (JSGlobalObject* globalObject, CallFr
+
+    JSObject* obj = asObject(value);
+
+    JSValue rawSize = obj->get(globalObject, vm.propertyNames->size);
+    RETURN_IF_EXCEPTION(scope, 0);
+
+    double numSize = rawSize.toNumber(globalObject);
+    RETURN_IF_EXCEPTION(scope, 0);
+
+    if (std::isnan(numSize)) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Set operation expects first argument to have non-NaN 'size' property"_s);
+
+    double intOrInfSize = jsNumber(numSize).toIntegerOrInfinity(globalObject);
+
+    if (intOrInfSize < 0) [[unlikely]]
+        return throwVMRangeError(globalObject, scope, "Set operation expects first argument to have non-negative 'size' property"_s);
+
+    if (std::isinf(intOrInfSize)) [[unlikely]]
+        return std::numeric_limits<uint32_t>::max();
+    return static_cast<uint32_t>(intOrInfSize);
+}
+
+static EncodedJSValue fastSetIntersection(JSGlobalObject* globalObject, JSSet* thisSet, JSSet* otherSet)
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSSet* result = JSSet::create(vm, globalObject->setStructure());
+
+    JSSet* sourceSet = thisSet->size() <= otherSet->size() ? thisSet : otherSet;
+    JSSet* targetSet = thisSet->size() <= otherSet->size() ? otherSet : thisSet;
+
+    JSCell* sourceStorageCell = sourceSet->storageOrSentinel(vm);
+    if (sourceStorageCell == vm.orderedHashTableSentinel())
+        return JSValue::encode(result);
+
+    JSSet::Storage& sourceStorage = *jsCast<JSSet::Storage*>(sourceStorageCell);
+    JSSet::Helper::Entry entry = 0;
+
+    while (true) {
+        sourceStorageCell = JSSet::Helper::nextAndUpdateIterationEntry(vm, sourceStorage, entry);
+        if (sourceStorageCell == vm.orderedHashTableSentinel())
+            break;
+
+        JSSet::Storage& currentStorage = *jsCast<JSSet::Storage*>(sourceStorageCell);
+        entry = JSSet::Helper::iterationEntry(currentStorage) + 1;
+        JSValue entryKey = JSSet::Helper::getIterationEntryKey(currentStorage);
+
+        if (targetSet->has(globalObject, entryKey)) {
+            result->add(globalObject, entryKey);
+            RETURN_IF_EXCEPTION(scope, JSValue::encode(result));
+        }
+
+        sourceStorage = currentStorage;
+    }
+    return JSValue::encode(result);
+}
+
+JSC_DEFINE_HOST_FUNCTION(setProtoFuncIntersection, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSSet* thisSet = getSet(globalObject, callFrame->thisValue());
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSValue otherValue = callFrame->argument(0);
+
+    if (otherValue.isObject() && setPrimordialWatchpointIsValid(vm, asObject(otherValue))) [[likely]] {
+        if (auto* otherSet = jsDynamicCast<JSSet*>(otherValue.asCell())) [[likely]] {
+            scope.release();
+            return fastSetIntersection(globalObject, thisSet, otherSet);
+        }
+    }
```

Comment:
I think we should first check `auto* otherSet = jsDynamicCast<JSSet*>(otherValue.asCell())`